### PR TITLE
WIP: sticking string length in the RTS

### PR DESCRIPTION
**Pull request by david-christiansen**
This is a WIP pull request so that the skilled C hackers can get critique in early. It's not done yet!

The idea is that an explicit length field means we're not dependent on null-termination. This has, in my view, two benefits:
1. Less chance of accidental buffer overruns from malformed UTF-8 strings
2. It becomes possible to use \0 inside of strings, giving the type `String` the semantics "sequence of code points" rather than "sequence of nonzero code points" which is at least elegant and possibly even useful.

A null terminator is retained to avoid marshalling issues or excess copying when calling C functions.

This pull request does the following:
- [X] Add a length field to the runtime representation of strings, tracking their size in bytes
- [X] Refactor string allocations to do all object initialization in one place, to help avoid mistakes
- [X] Modify the UTF-8 functions to not access memory outside the size size of the string
- [X] Crash with informative messages on partial string operations (like `strTail ""`)
- [X] Validate the output of primitive conversions, like `prim__floatToStr` to make sure that library code is returning what we expect
- [ ] Stop using C string functions that assume null-termination (partially complete)
- [ ] Check for `NULL` return from all `malloc` calls (partially complete)
- [X] Pass the test suite
- [ ] Validate that strings from calls to C code are actually real UTF-8 (I may punt on this or implement a poor version of it until we get a proper Unicode library integrated)
- [X] Validate that purported characters (represented as `Int` under the hood) are actually in the Unicode range before consing onto a string
- [X] Add `assert`s about the byte size invariants
- [ ] Make the C codegen actually allow `'\0'` in the middle of Idris strings, and display them appropriately using the print functions in the library
- [ ] Factor out checks for `NULL` after `malloc` into a function

Potential things to do:
1. Also cache the length in characters (would make some things more elegant, but adds overhead and complexity and may not be worth it)
2. More thorough Unicode testing

### CODE REVIEW

**Comment by david-christiansen** [comment]
Comments, critique, and questions are most certainly welcome!

**Comment by david-christiansen** [comment]
And please, even when the tests pass, nobody click "Merge" - when it's ready and I've got buy-in from the right people, I'll do so. Thanks!

**Comment by david-christiansen** [comment]
@edwinb, I'd like to hear what you think, at least about the idea.

@Melvar, @mbj, @daira, @melted: you were very helpful last time I poked at some C - if you have time and energy, I'd appreciate further feedback!

**Comment by mbj** [comment]
I really like the intention of this PR. The cstring data format and its in-line signalling are a major bug class source. Avoiding this construct eliminates a bug class. So big +1.

I do not have the time to do a dense review. Nor followup with the old discussions, been sick the last weeks and need to cure (AKA sleeping). BTW if anyone of you goes to StrangeLoop: I'll be in the Idris workshop by @edwinb would be cool to meet more people from here in person.

**Comment by melted** [comment]
I'll go through it.

> And please, even when the tests pass, nobody click "Merge" - when it's ready and I've got buy-in from the right people, I'll do so. Thanks!

I wonder if there's any way to disable the merge button temporarily, except by introducing an intentional merge conflict.

**Comment by mbj** [comment]
> I wonder if there's any way to disable the merge button temporarily, except by introducing an intentional merge conflict.

Githup repos in organizations have the concept of "protected branches", these right now prevent merges when *any* configured integration is red (example Travis).

There could be a custom integration that adds some kind of "review overlay", it would only report green when all assigned people did a special comment like `[agreeed]` etc. Its true such a system requires coding + external state keeping. But maybe such an overlay bot already exists.

File: rts/idris_gmp.c [hunk 1]
```diff
@@ -374,11 +375,37 @@ VAL idris_castFloatBig(VM* vm, VAL f) {
...
+
+
+    // Compute the appropriate size in bytes for a C string holding the result
+    size_t max_len = mpz_sizeinbase(GETMPZ(GETBIG(vm, i)), 10) + 2;
+    char *temp = (char *)malloc(max_len);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate %lu bytes for bigint as string", (unsigned long)max_len);
+        exit(EXIT_FAILURE);
+    }
+    memset(temp, '\0', max_len);
```
Comment (melted) [inline rts/idris_gmp.c]: can use calloc to allocate zeroed memory

File: rts/idris_gmp.c [hunk 2]
```diff
@@ -374,11 +375,37 @@ VAL idris_castFloatBig(VM* vm, VAL f) {
...
+    mpz_get_str(temp, 10, GETMPZ(GETBIG(vm, i)));
+
+    unsigned int j;
+    for (j = 0; j < max_len; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j])) && temp[j] != '-') {
```
Comment (melted) [inline rts/idris_gmp.c]: Why can't we trust that we get digits from GMP or minigmp?

File: rts/idris_rts.c [hunk 3]
```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
         fprintf(stderr, "Fatal Error: ClosureType %d, not an integer type", ty);
         exit(EXIT_FAILURE);
     }
 
-    SETTY(cl, STRING);
-    return cl;
+    // avoid invalid UTF-8: all ASCII digits are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < 21; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j]))) {
```
Comment (melted) [inline rts/idris_rts.c]: this section does exactly the same as the one above, should make a function validateNumber out of it, though I suspect the other place s uneccesary. Also, that 21 shouldn't be hardcoded like that. Magic constants should be defined in one place with a comment about what they represent.

File: rts/idris_rts.c [hunk 4]
```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
-    return cl;
+    char temp[32] = { 0 };
+    // snprintf will put a null at the end, so 31 chars are available for the float
+    snprintf(temp, 32, "%.16g", GETFLOAT(i));
+
+    // avoid invalid UTF-8: all ASCII digits, plus period, plus, and
+    // minus, are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < 21; j++) {
+        if (temp[j] != '\0' &&
```
Comment (melted) [inline rts/idris_rts.c]: Magic constants, also mixup of 21 and 32

File: rts/idris_rts.c [hunk 5]
```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
 }
 
 VAL idris_castStrFloat(VM* vm, VAL i) {
-    return MKFLOAT(vm, strtod(GETSTR(i), NULL));
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
+    char *temp = (char *) malloc(str.size + 1);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate temp buffer of size %lu for int->string conversion", str.size);
```
Comment (melted) [inline rts/idris_rts.c]: I don't know about the usability of checking for failure on malloc, everybody has overcommit. Linux failure mode is to freak out when you try to use the memory. It will never fail to allocate memory as long as there is address space.

File: rts/idris_rts.c [hunk 6]
```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
+
+
+    return MKSTR(vm, (utf8_byte *)temp);
 }
 
 VAL idris_castStrFloat(VM* vm, VAL i) {
-    return MKFLOAT(vm, strtod(GETSTR(i), NULL));
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
```
Comment (melted) [inline rts/idris_rts.c]: hey, don't we already have null termination?

File: rts/idris_rts.c [hunk 7]
```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
+        exit(EXIT_FAILURE);
+    }
+    // Then grab the UTF8 bytes out of the Idris string. We don't have
+    // to worry about nulls in the middle because we'll later check to
+    // see if the whole string was used.
+    memcpy(temp, str.str_data, str.size);
+    temp[str.size] = '\0';
+
+    char *end;
+    double v = strtod(temp, &end);
```
Comment (melted) [inline rts/idris_rts.c]: I like scanf better, say `scanf("%lf", &v)`, it returns 1 (for nbr of read values) on success.

File: rts/idris_rts.c [hunk 8]
```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
 }
 
 VAL idris_castStrInt(VM* vm, VAL i) {
     char *end;
-    i_int v = strtol(GETSTR(i), &end, 10);
-    if (*end == '\0' || *end == '\n' || *end == '\r')
-        return MKINT(v);
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
```
Comment (melted) [inline rts/idris_rts.c]: wasn't it nullterminated in the VAL?

File: rts/idris_rts.c [hunk 9]
```diff
@@ -556,62 +673,83 @@ VAL idris_strTail(VM* vm, VAL str) {
             root = root->info.str_offset->str;
         }
 
-        cl->info.str_offset->str = root;
-        cl->info.str_offset->offset = offset+idris_utf8_charlen(GETSTR(str));
+        cl -> info.str_offset -> str = root;
+        cl -> info.str_offset -> offset = offset + idris_utf8_charlen(string);
```
Comment (melted) [inline rts/idris_rts.c]: no space around pointer arrows.

File: rts/idris_rts.c [hunk 10]
```diff
@@ -556,62 +673,83 @@ VAL idris_strTail(VM* vm, VAL str) {
...
         return cl;
     } else {
-        char* nstr = GETSTR(str);
-        return MKSTR(vm, nstr+idris_utf8_charlen(nstr));
+        SizedString nstr = GETSTR(str);
+        unsigned int tailIdx = idris_utf8_charlen(nstr);
+        assert(tailIdx <= nstr.size);
+        Closure *cl = idris_allocate_string(nstr.size - tailIdx, 0);
+        // Use memcpy here instead of MKSTR because MKSTR assumes
+        // C-style strings, and 0 is a sensible Unicode character that
```
Comment (melted) [inline rts/idris_rts.c]: wouldn't it be reasonable then that MKSTR takes a length argument?

File: rts/idris_utf8.h [hunk 11]
```diff
@@ -1,24 +1,38 @@
...
+#include <stdint.h>
+#include <stdlib.h>
+
 /* Various functions for dealing with UTF8 encoding. These are probably
    not very efficient (and I'm (EB) making no guarantees about their
    correctness.) Nevertheless, they mean that we can treat Strings as
    UFT8. Patches welcome :). */
 
+typedef uint8_t utf8_byte;
+
```
Comment (melted) [inline rts/idris_utf8.h]: not a fan of this typedef. uint8_t is always the correct type for it, and nobody will be confused by utf8 being represented by an array of uint8_t.

File: rts/idris_utf8.h [hunk 12]
```diff
@@ -1,24 +1,38 @@
...
+// Pascal-style strings with their lengths.
+typedef struct {
+    size_t size;
+    utf8_byte* str_data;
+} SizedString;
```
Comment (melted) [inline rts/idris_utf8.h]: Could save some indirection by appending str_data as a zero-length array. It won't be computable at compile time so lose sizeof, but the length is in there so easy to substitute.

**Comment by melted** [comment]
MUST NOT PRESS BIG GREEN BUTTON... MUST NOT PRESS BIG GREEN BUTTON...

File: rts/idris_utf8.h (hunk 11 above)
Comment (david-christiansen) [inline rts/idris_utf8.h]: The reason I did it was to document the intention at use sites - that this particular `uint8` is intended as part of a string. Is that not reasonable?

File: rts/idris_gmp.c (hunk 2 above)
Comment (david-christiansen) [inline rts/idris_gmp.c]: We probably can. Really, my goal is to ensure that all in-memory strings are valid UTF-8, so I want to do this kind of validation for each source of strings.

File: rts/idris_rts.c (hunk 5 above)
Comment (mbj) [inline rts/idris_rts.c]: On embedded devices checking for malloc fails is critical. If idris ever wants to support such environments it needs to check each malloc return. Ideally idris uses a centralized `idris_malloc` or something more sophisticated to centralize detection and handling of the issue.

File: rts/idris_gmp.c (hunk 1 above)
Comment (david-christiansen) [inline rts/idris_gmp.c]: Thanks, fixed.

File: rts/idris_gmp.c (hunk 2 above)
Comment (david-christiansen) [inline rts/idris_gmp.c]: This might go wrong, for example, if we accidentally specified the wrong base.

File: rts/idris_rts.c (hunk 10 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: `MKSTR` is really responsible for turning actual C strings into Idris RTS strings. It's used various places in the codegen, along with `MKINT` and friends. It might be feasible to plumb lengths through that, however, and that will probably be necessary in the end to allow `'\0'` as part of a string.

File: rts/idris_rts.c (hunk 5 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: I think that, for running on very small systems, we'd likely need to do the RTS differently, and these malloc fail checks are the least of our worries! But I still prefer to do it, even if it's highly unlikely to happen on Linux, because it should increase the portability to new systems that we haven't anticipated.

Comment (melted) [inline rts/idris_rts.c]: As it exits on allocation failure it could be made into a function and there wouldn't be a need to handle the error on every call site.

----- Ursprungligt meddelande -----
Från: "David Christiansen" <notifications@github.com>
Skickat: ‎2015-‎09-‎18 08:40
Till: "idris-lang/Idris-dev" <Idris-dev@noreply.github.com>
Kopia: "Niklas Larsson" <metaniklas@gmail.com>
Ämne: Re: [Idris-dev] WIP: sticking string length in the RTS (#2628)

In rts/idris_rts.c:
>  }
>  
>  VAL idris_castStrFloat(VM* vm, VAL i) {
> -    return MKFLOAT(vm, strtod(GETSTR(i), NULL));
> +    SizedString str = GETSTR(i);
> +
> +    // First, we need to alloc a C string that's long enough (to get null-termination)
> +    char *temp = (char *) malloc(str.size + 1);
> +    if (temp == NULL) {
> +        fprintf(stderr, "Couldn't allocate temp buffer of size %lu for int->string conversion", str.size);
I think that, for running on very small systems, we'd likely need to do the RTS differently, and these malloc fail checks are the least of our worries! But I still prefer to do it, even if it's highly unlikely to happen on Linux, because it should increase the portability to new systems that we haven't anticipated.

Comment (david-christiansen) [inline rts/idris_rts.c]: That's certainly the case.

File: rts/idris_utf8.h (hunk 12 above)
Comment (david-christiansen) [inline rts/idris_utf8.h]: The cost of this is that the code gets more complex. Given that the `str_data` array is always allocated immediately adjacent to the `SizedString`, it should be in the same cache line. Would we really win much there to make it worth the more convoluted code?

File: rts/idris_rts.c (hunk 6 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: You're right! This is obsolete - I was going to get rid off the null termination, but that didn't happen here. I'll fix it!

File: rts/idris_rts.c (hunk 7 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: That won't return a pointer to the end of what was read, though, will it? We need that to check if the whole string was consumed.

File: rts/idris_rts.c (hunk 4 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: Def. needs fixing. Thanks!

File: rts/idris_rts.c (hunk 3 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: I was looking at validating it in a function, but the different types admit different characters in their string representations. Making that into a nice function would make things crazier, in my view. I did remove the magic numbers though.

File: rts/idris_rts.c (hunk 4 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: OK, magic numbers replaced.

File: rts/idris_rts.c (hunk 5 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: The downside is that it fails to allow for custom messages - but I suppose I could make a macro that passes `__LINE__` to make it so we can find where it happened.

File: rts/idris_rts.c (hunk 8 above)
Comment (david-christiansen) [inline rts/idris_rts.c]: it is, but the size field is the size of the actual data, not including the `'\0'` at the end for C

File: rts/idris_rts.c (hunk 5 above)
Comment (melted) [inline rts/idris_rts.c]: I don'rt see why you'd need a custom message if malloc fails. It's a system thing, the only thing that could be of interest would be the size of the allocation. It's an extreme corner case anyhow, and it's very unlikely the system will live long enough to print anything if small allocs fail.

Many programs don't bother to handle alloc failure at all (and OpenBSD's malloc just aborts the program). It's an extreme corner case, the least we can do is to handle it in one place and not sprinkle error handling for it all over.

File: rts/idris_rts.c (hunk 3 above)
Comment (daira) [inline rts/idris_rts.c]: I came to the same conclusion as @melted -- we do similar validation in about 4 places IIRC, so it is definitely worth factoring out. It is ok if the set of characters accepted is an overapproximation as long as it is a subset of US-ASCII.

File: rts/idris_rts.c (hunk 5 above)
Comment (daira) [inline rts/idris_rts.c]: +1 for centralizing checks for malloc failures.

File: rts/idris_rts.c [hunk 13]
```diff
@@ -411,102 +435,193 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
+    // see if the whole string was used.
+    memcpy(temp, str.str_data, str.size);
+    temp[str.size] = '\0';
+
+    char *end;
+    double v = strtod(temp, &end);
+    Closure* answer;
+    // If we used the whole string, return the double, otherwise
+    // return 0.0. This is the meaning of the primitive in the
+    // evaluator.
```
Comment (daira) [inline rts/idris_rts.c]: I think this is error-prone and should be reconsidered.

**Comment by david-christiansen** [comment]
Thanks for the feedback, @daira! As I get some free time where I need to de-stress from PhD writing, I'll go through and take care of details. Appreciate it!

File: rts/idris_rts.c (hunk 13 above)
Comment (Melvar) [inline rts/idris_rts.c]: It is, however, the current semantic of the primitive. Changing that is a separate issue.

Comment (daira) [inline rts/idris_rts.c]: Filed #2758.

**Comment by ahmadsalim** [comment]
@david-christiansen Are you still planning to complete this when you have time?

**Comment by ahmadsalim** [comment]
@david-christiansen Or let me rephrase, would it be possible to partition tasks into a `nice-to-have` and a `necessary-to-complete` section, and then finishing up the `necessary-to-complete` parts sooner, while converting the others to issues?

I think this feature is useful, and it would be sad if this work was never finished because of some `nice-to-have` requirements like allowing `\0` mid string.

File: rts/idris_gmp.c [hunk 14]
```diff
@@ -374,11 +375,36 @@ VAL idris_castFloatBig(VM* vm, VAL f) {
 }
 
 VAL idris_castStrBig(VM* vm, VAL i) {
-    return MKBIGC(vm, GETSTR(i));
+    // TODO: validate that the string is actually an integer
```
Comment (ericqweinstein) [inline rts/idris_gmp.c]: Can we use `isnumber` for this? (`$ man isnumber`)

File: rts/idris_rts.c [hunk 15]
```diff
@@ -411,102 +435,193 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
+    // to worry about nulls in the middle because we'll later check to
+    // see if the whole string was used.
+    memcpy(temp, str.str_data, str.size);
+    temp[str.size] = '\0';
+
+    i_int v = strtol(temp, &end, 10);
+    Closure* answer;
+    // If we used the whole string, return the int, otherwise return 0. This is
+    // the meaning of the primitive in the evaluator.
+    if (end == temp + str.size)
```
Comment (ericqweinstein) [inline rts/idris_rts.c]: V. small (and possibly paranoia-related and not in keeping with the style guide), but is it beneficial to include braces here, even for the single line? (Here I'm thinking of [#gotofail](https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/).)

```c
if (end == temp + str.size) {
  answer = MKINT(v);
} else {
  answer = MKINT(0);
}

File: rts/idris_rts.c [hunk 16]
```diff
@@ -411,102 +435,193 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
...
+    char *end;
+    double v = strtod(temp, &end);
+    Closure* answer;
+    // If we used the whole string, return the double, otherwise
+    // return 0.0. This is the meaning of the primitive in the
+    // evaluator.
+    if (end == temp + str.size)
+        answer = MKFLOAT(vm, v);
+    else
+        answer = MKFLOAT(vm, 0.0);
```
Comment (ericqweinstein) [inline rts/idris_rts.c]: Ditto above q re: braces

File: rts/idris_rts.c [hunk 17]
```diff
@@ -540,9 +658,15 @@ VAL MKSTROFFc(VM* vm, StrOffset* off) {
...
+    SizedString string = GETSTR(str);
+    // Crash with a message when taking the tail of an empty string
+    if (string.size < 1) {
+        fprintf(stderr, "Tried to take the tail of an empty string.\n");
+        exit(EXIT_FAILURE);
+    } else if (space(vm, sizeof(Closure) + sizeof(StrOffset))) {
     // If there's no room, just copy the string, or we'll have a problem after
     // gc moves str
-    if (space(vm, sizeof(Closure) + sizeof(StrOffset))) {
+//        printf("space\n");
```
Comment (ericqweinstein) [inline rts/idris_rts.c]: Should we remove this commented-out line?

File: rts/idris_utf8.c [hunk 18]
```diff
@@ -3,17 +3,25 @@
...
-     if ((s[i] & 0xc0) != 0x80) j++;
+size_t idris_utf8_unitlen(const utf8_byte* s) {
+    return strlen((const char*) s);
+}
+
+unsigned int idris_utf8_strlen(SizedString s) {
+   unsigned int i = 0, j = 0;
+   while (i < s.size) {
+     if ((s.str_data[i] & 0xc0) != 0x80) j++;
      i++;
```
Comment (ericqweinstein) [inline rts/idris_utf8.c]: Ditto above q re: braces; might improve readability here?

**Comment by david-christiansen** [comment]
@ahmadsalim I would like to finish this!  If you want to partition bits up, let me know, but otherwise I just need to find a weekend afternoon without other plans.

**Comment by ericqweinstein** [comment]
I'm also happy to help out here!

**Comment by david-christiansen** [comment]
@ericqweinstein Thanks! Please feel free - C is far from my first language, and it's easy to get things wrong, so I appreciate your help!

**Comment by ahmadsalim** [comment]
@ericqweinstein Did you ever take a look at this issue?

**Comment by ericqweinstein** [comment]
@ahmadsalim I haven't had time to look yet, but could take a look toward the end of this week

**Comment by ahmadsalim** [comment]
@ericqweinstein That would be very nice if possible! Thanks.

**Comment by ericqweinstein** [comment]
@ahmadsalim Sure thing! Sorry for the delay, busy month/semester thus far. Three-day weekend though, so I'll take a look by Monday night!

**Comment by ahmadsalim** [comment]
@ericqweinstein That sounds great :). I know that you (and many other people) are busy, which is why I try to nudge a bit once in a while.

**Comment by ericqweinstein** [comment]
@ahmadsalim Just wanted to let you know that I'm still interested in this, but due to the work i've got on my plate now, I don't know when I'll be able to get to it—I'll update here if/when things change!

**Comment by ahmadsalim** [comment]
@ericqweinstein OK, I understand. Thanks for looking into it anyway, and please do tell us if you ever come to finish it :)

**Comment by ahmadsalim** [comment]
@david-christiansen Any ideas what we should do about this PR? It has been open for a long time :).

**Comment by ahmadsalim** [comment]
I will close the PR since it has been opened for a long time and no one seems to have had time to fix the issues. If people are interested in fixing the issues please reopen the PR with the required code updates.
